fem/convergence.hpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_CONVERGENCE
+#define MFEM_CONVERGENCE
+
+#include "../linalg/linalg.hpp"
+#include "gridfunc.hpp"
+#ifdef MFEM_USE_MPI
+#include "pgridfunc.hpp"
+#endif
+
+namespace mfem
+{
+
+/** @brief Class to compute error and convergence rates.
+    It supports H1, H(curl) (ND elements), H(div) (RT elements) and L2 (DG). */
+class Convergence
+{
+private:
+   // counters for solutions/derivatives
+   int counter=0;
+   int dcounter=0;
+   int fcounter=0;
+
+   // space continuity type
+   int cont_type=-1;
+
+   // printing flag for helpful for MPI calls
+   int print_flag=1;
+
+   // exact solution and derivatives
+   double CoeffNorm;
+   double CoeffDNorm;
+
+   // Arrays to store error/rates
+   Array<double> L2Errors, DGFaceErrors, DErrors, EnErrors;
+   Array<double> L2Rates, DGFaceRates, DRates, EnRates;
+   Array<int> ndofs;
+
+   void AddL2Error(GridFunction * gf, Coefficient * u, VectorCoefficient * U);
+   void AddGf(GridFunction * gf, Coefficient * u, VectorCoefficient * grad,
+              Coefficient * ell_coeff, double Nu);
+   void AddGf(GridFunction * gf, VectorCoefficient * u,
+              VectorCoefficient * curl, Coefficient * div);
+
+   // returns the exact solution/grad/div/curl norm
+   double GetNorm(GridFunction * gf, Coefficient * u, VectorCoefficient * U);
+
+public:
+
+   // Clear any internal data
+   void Reset();
+
+   // Add H1/L2 GridFunction,
+   // the exact solution and possibly
+   // its gradient and/or DG face jumps parameters
+   void AddGridFunction(GridFunction * gf, Coefficient * u,
+                        VectorCoefficient * grad=NULL,
+                        Coefficient * ell_coeff=NULL, double Nu=1.0)
+   {
+      AddGf(gf, u, grad, ell_coeff,Nu);
+   }
+
+   // Add H(curl)/H(div) GridFunction,
+   // and the exact solution
+   void AddGridFunction(GridFunction * gf, VectorCoefficient * u)
+   {
+      AddGf(gf,u, nullptr, nullptr);
+   }
+
+   // Add H(curl) GridFunction,
+   // the exact solution and its curl
+   void AddGridFunction(GridFunction * gf, VectorCoefficient * u,
```

Comment:
The reason I preferred to do it this way is because the user would just call one function to register the GridFunction and the exact solution, independently of the space. I am okay either way though...

Reply:
@tzanio @v-dobrev , any thoughts? I'm equally OK with either option.

Reply:
I think it will be more clear to change the names to `AddL2GridFunction`, `AddH1GridFunction`, `AddHcurlGridFunction` and  `AddHdivGridFunction`. There is already a switch for the different cases, and the different names can actually help with readability.

Reply:
👍